Metal capture support for wrapped MTLCommandBuffer

Zorro666 opened the pull request:
## Description

Added wrapped `MTLCommandBuffer`.
Implemented capture serialisation for:
`MTLCommandQueue::commandBuffer`
`MTLCommandBuffer::presentDrawable`
`MTLCommandBuffer::commit`

The command buffer record contains `cmdInfo` (`CmdBufferRecordingInfo`).

The current record, chunk hierarchy summary:
- `MTLCommandQueue` Record is set to be the parent of the `MTLCommandBuffer` Record.
    - This is done during `commit` when it is known the command buffer will be submitted and included in the capture. Also during `commit` the `MTLCommandQueue` record is marked as frame referenced.
- `MTLCommandBuffer` Record has the command buffer chunks added to it ie. 
    - `create RenderCommandEncoder` (in a future PR), `present`, `commit`.
-  `MTLCommandBuffer` Record has the `RenderCommandEncoder` chunks added to it ie. in future PRs
    - `setRenderPipelineState`, `setVertexBuffer`, `setFragmentBuffer`, `setViewport`, `drawPrimitives`, `endEncoding`

In a future PR if the frame is being captured then `commit` adds the command buffer record to an array of command buffer records in the device. The array of command buffer records then has its chunks serialized into the capture. In the future, the device will also contain a submission order array to support serializing the command buffers in the driver submission order. By default, the driver submission order is the `commit` API order, except that the `queue` API can reserve a submission slot before calling `commit`.

Once `commit` is called then `MTLCommandBuffers` are [immutable](https://developer.apple.com/documentation/metal/mtlcommandbuffer?language=objc) 

> After a command buffer has been committed for execution, the only valid operations on the command buffer are to wait for it to be scheduled or completed (using synchronous calls or handler blocks) and to check the status of the command buffer execution.

The overall goals of the RenderDoc Metal capture Records and Chunks are captured [here](https://github.com/Zorro666/renderdoc/projects/1#card-80073363) for discussion and development.

Zorro666 committed:
Added wrapped MTLCommandBuffer

Implemented capture serialisation for 
MTLCommandQueue::commandBuffer
MTLCommandBuffer::presentDrawable
MTLCommandBuffer::commit

Zorro666 committed:
De-serialise ResourceId default resource to NULL

Used during structured exporting

Zorro666 commented on renderdoc/driver/metal/metal_command_buffer.cpp:
```diff
@@ -0,0 +1,122 @@
+      Serialise_commit(ser);
+      chunk = scope.Get();
+    }
+    GetResourceManager()->MarkResourceFrameReferenced(GetResID(m_WrappedMTLCommandQueue),
```
The marking of the parent wrapped command queue might change in the future.
See [here](https://github.com/Zorro666/renderdoc/issues/58) for the thoughts and goals about the chunks for the creation of Permanent Resources.

Zorro666 replied:
In a future PR, this is likely to be changed to only mark the command queue as referenced during an active capture.
```
bool capframe = IsActiveCapturing(m_State);
if(capframe)
{
    GetResourceManager()->MarkResourceFrameReferenced(GetResID(m_WrappedMTLCommandQueue), eFrameRef_Read);
}
```

baldurk replied:
I'm not sure this mark reference here is necessary or the best fit for what I'd expect. If the queue is the parent of a command buffer then its creation record will be included regardless as long as the command buffer is marked as referenced. Generally speaking marking a resource referenced should happen when it is explicitly used, so e.g. if a queue operation happened mid-frame then that would mark the queue as referenced. It would feel more natural to mark the command buffer itself as referenced since this function is the one which submits the work in it, though since command buffers usually have special lifetime management it's maybe more likely that neither are needed.

Also yes, resources should usually only be marked as referenced during active capturing.

Zorro666 commented on renderdoc/driver/metal/metal_types.cpp:
```diff
@@ -49,16 +50,19 @@ void DoSerialiseViaResourceId(SerialiserType &ser, type &el)
-  if(ser.IsReading() && rm && !IsStructuredExporting(rm->GetState()))
+  if(ser.IsReading())
   {
     el = NULL;
```
Found this problem when doing the conversion from rdc -> XML. 
A serialized NS::String* was not set to NULL causing a crash.

baldurk replied:
Why was an `NS::String*` going through this path being serialised as an Id?

baldurk commented on renderdoc/driver/metal/metal_command_buffer.h:
```diff
@@ -0,0 +1,63 @@
+  WrappedMTLCommandQueue *m_WrappedMTLCommandQueue;
+};
+
+inline MTL::CommandQueue *WrappedMTLCommandBuffer::GetObjCBridgeMTLCommandQueue()
```
These types of small inline functions can be written directly in the class definition rather than separately but still in the header.

baldurk commented on renderdoc/driver/metal/metal_command_queue.cpp:
```diff
@@ -31,3 +32,58 @@ WrappedMTLCommandQueue::WrappedMTLCommandQueue(MTL::CommandQueue *realMTLCommand
+    bufferRecord->cmdInfo = new MetalResources::CmdBufferRecordingInfo();
+    bufferRecord->cmdInfo->queue = this;
+    bufferRecord->cmdInfo->device = m_WrappedMTLDevice;
+    bufferRecord->cmdInfo->present = false;
```
I'd prefer to set default values like this in the class definition or default constructor (as appropriate). Doing it here means it's easier to forget to set them somewhere else if this class needs to be initialised somewhere else in future.

baldurk commented on renderdoc/driver/metal/metal_resources.h:
```diff
@@ -127,6 +128,28 @@ struct UnwrapHelper
 METALCPP_WRAPPED_PROTOCOLS(WRAPPED_TYPE_HELPERS)
 #undef WRAPPED_TYPE_HELPERS
 
+namespace MetalResources
```
What's the motivation for this namespace? The metal is implicit given the location of this code, and 'resources' is a bit generic and possibly misleading because the class in here (and presumably others that will follow) are _not_ resources themselves but rather resource-specific cold data.

baldurk commented on renderdoc/driver/metal/metal_command_buffer.cpp:
```diff
@@ -0,0 +1,122 @@
+                                                      eFrameRef_Read);
+    MetalResourceRecord *queueRecord = GetRecord(m_WrappedMTLCommandQueue);
+    MetalResourceRecord *record = GetRecord(this);
+    record->AddParent(queueRecord);
```
Calling `AddParent` should only happen at creation time or other initialisation calls, since it represents a dependency for creation. In this case command buffers created directly from queues so the parent relationship can be added immediately.

Zorro666 committed:
PR feedback for CmdBufferRecordingInfo

Removed namespace MetalResources scope
Rename it to MetalCmdBufferRecordingInfo to remove conflict with Vulkan struct of the same name.
Mark MetalCmdBufferRecordingInfo default constructor as deleted
Added MetalCmdBufferRecordingInfo constructor which takes WrappedMTLCommandQueue* and initializes member variables to sensible defaults.
Removed "WrappedMTLDevice *device;" member variable.

Zorro666 committed:
PR Feedback for WrappedMTLCommandBuffer

Moved small inline methods to be inside the class declaration instead of outside

Zorro666 committed:
MTLCommandBuffer::commit record/chunk tweak

Do not set the MTLCommandQueue record as a parent of the MTLCommandBuffer record.
Only mark the MTLComamndQueue record as frame referenced when active capturing.

Zorro666 committed:
PR Feedback MTLCommandBuffer::commit record tweak

MarkResourceFrameReferenced() on the buffer record instead of the resource manager
Transfer the frame references on the buffer record to the resource manager when the buffer is committed.
